wire-ios-request-strategy/Sources/Message Sending/MessageInfoExtractor.swift

```diff
@@ -0,0 +1,200 @@
+            // We do not want to send pushes for delivery receipts.
+            nativePush: !message.hasConfirmation
+        )
+    }
+
+    private func selfClientID() async throws -> String {
+        let selfClientID = await context.perform {
+            ZMUser.selfUser(in: context).selfClient()?.remoteIdentifier
+        }
+        guard let id = selfClientID else { throw MessageInfoExtractorError.missingValidSelfClient }
+        return id
+    }
+
+    private func recipients(for message: GenericMessage, selfUser: ZMUser, in conversation: ZMConversation) async -> ([ZMUser: Set<UserClient>], MissingClientsStrategy) {
+        let (users, missingClientsStrategy) = await context.perform { message.recipientUsersForMessage(in: conversation, selfUser: selfUser) }
+
+        return (users, missingClientsStrategy)
+    }
+
+    private func listOfClients(for recipients: [ZMUser: Set<UserClient>], selfDomain: String, selfClientID: String) async -> MessageInfo.ClientList {
+
+        let recipientsByDomain = await context.perform {
+            Dictionary(grouping: recipients) { element -> String in
+                element.key.domain ?? selfDomain
+                // is there really a need to keep selfDomain as backup values
+            }
+        }
+
+        var qualifiedUserEntries = MessageInfo.ClientList()
+        for (domain, recipients) in recipientsByDomain {
+
+            var userEntries = [MessageInfo.UserID: [UserClientData]]()
+            for (user, clients) in recipients {
+                guard let userId = await context.perform({
+                    !user.isAccountDeleted ? user.remoteIdentifier : nil
+                }) else { continue }
+
+                let userClientDatas = await userClientDatas(selfClientID: selfClientID, userClients: clients)
+                if !userClientDatas.isEmpty {
+                    userEntries[userId] = userClientDatas
+                }
+            }
+            qualifiedUserEntries[domain] = userEntries
+        }
+        return qualifiedUserEntries
+    }
+
+    private func userClientDatas(selfClientID: String, userClients: Set<UserClient>) async -> [UserClientData] {
+
+        return await context.perform {
+            userClients.compactMap {
+                guard let sessionID = $0.proteusSessionID,
+                      $0.remoteIdentifier != selfClientID else {
+                    // skips self client session
+                    WireLogger.proteus.warn("skips cliend id: \(String(describing: $0.remoteIdentifier)), proteusSession id: \(String(describing: $0.proteusSessionID))")
+                    return nil
+                }
+
+                guard !$0.failedToEstablishSession else {
+                    let data = ZMFailedToCreateEncryptedMessagePayloadString.data(using: .utf8)!
+                    WireLogger.proteus.error("Failed to encrypt payload: session is not established with client: \(String(describing: $0.remoteIdentifier))")
```

Comment:
```suggestion
                    WireLogger.proteus.error("Failed to encrypt payload: session couldn't be established with client: \(String(describing: $0.remoteIdentifier))")
```

It's not that the session is not established, it's that we tried to establish it and we failed.